CONFIGURING.md
## List of options
This is a list of all available options. The configuration file is written in `TOML` format and must be located in `$XDG_CONFIG_HOME/lyrics/config.toml` (usually `~/.config/lyrics/config.toml`). Each of the following sub-sections should also be a section in the config file.

Each subsection below corresponds to a section in the config file.

An example config file can be found [here](/config.toml)

---

### General `[general]`

| Name                | Description                                                                                                         | Accepted Values           | Default  |
| ------------------- | ------------------------------------------------------------------------------------------------------------------- | ------------------------- | -------- |
| `program_name`      | The name used to identify the program (e.g., in notifications).                                                     | Any `string`              | `muse`   |
| `check_for_updates` | Notifies when program version is different from latest version on GitHub.                                           | `true`, `false`           | `true`   |
| `display_mode`      | How lyrics should be displayed. `window` or `tui` are intended for normal use, `simple` is for scripting/debugging. | `window`, `tui`, `simple` | `simple` |

### Player `[player]`

| Name              | Description                                                                                                                   | Accepted Values  | Default |
| ----------------- | ----------------------------------------------------------------------------------------------------------------------------- | ---------------- | ------- |
| `player`          | The media player to connect to. Any string contained in the player's name suffices.                                           | Any `string`     | `"mpv"` |
| `position_offset` | Time offset (in seconds) applied to the current playback position.                                                            | Any `int`, in ms | `-0.52` |
| `step`            | The time interval (in seconds) used for polling/stepping through playback.                                                    | Any `int`, in ms | `0.3`   |
| `poll_interval`   | A cooldown period between song changes to prevent spamming the API. The default value does not interrupt the user experience. | Any `int`, in ms | `3000`  |
| `silence_timeout` | The maximum silence duration (in seconds) before lyrics are considered paused/stopped.                                        | Any `int`, in ms | `3`     |

### Display `[display]`

| Name         | Description                                     | Accepted values     | Default   |
| ------------ | ----------------------------------------------- | ------------------- | --------- |
| `foreground` | Text color used in display mode.                | Any hex color value | `#ffffff` |
| `background` | Background color used in display mode.          | Any hex color value | `#000000` |

### Display `[display]` / *Exclusive to `window` mode*

| Name            | Description                                                                  | Accepted values              | Default |
| --------------- | ---------------------------------------------------------------------------- | ---------------------------- | ------- |
| `font`          | Typeface used when rendering lyrics. Leave empty for fallback.               | Any `string`, use full paths | `""`    |
| `font_size`     | Font size for the lyrics displayed.                                          | Any `int`, in px             | `32`    |
| `window_x`      | Window's spawn position. May not work depending on the compositor.           | Any `int`, in px             | `410`   |
| `window_y`      | Window's spawn position. May not work depending on the compositor.           | Any `int`, in px             | `0`     |
| `window_width`  | Window's spawn dimensions. May not work depending on the compositor.         | Any `int`, in px             | `1100`  |
| `window_height` | Window's spawn dimensions. May not work depending on the compositor.         | Any `int`, in px             | `250`   |


internal/config/config.go
package config

import (
	"muse/internal/util"

	"github.com/BurntSushi/toml"
)

type Config struct {
	General  General  `toml:"general"`
	Player   Player   `toml:"player"`
	Fetching Fetching `toml:"fetching"`
	Display  Display  `toml:"display"`
}

type General struct {
	ProgramName string `toml:"program_name"`
	DisplayMode string `toml:"display_mode"`
	Updates     bool   `toml:"check_for_updates"`
}

type Player struct {
	Preferred      []string `toml:"preferred"`
	PositionOffset float32  `toml:"position_offset"` // in seconds
	PollInterval   int      `toml:"poll_interval"`   // in milliseconds
	SilenceTimeout int      `toml:"silence_timeout"` // in milliseconds
}

type Fetching struct {
}

type Display struct {
	FgColor  string `toml:"foreground"`
	BgColor  string `toml:"background"`
	Font     string `toml:"font"`
	FontSize int    `toml:"font_size"`

	WindowX int `toml:"window_x"`
	WindowY int `toml:"window_y"`
	WindowW int `toml:"window_width"`
	WindowH int `toml:"window_height"`
}

func Default() *Config {
	var c Config

	c.General.ProgramName = "muse"
	c.General.DisplayMode = "display"
	c.General.Updates = true

	c.Player.Preferred = []string{"tauon", "mpv", "spotify", "chromium"}
	c.Player.PositionOffset = -0.52
	c.Player.PollInterval = 250
	c.Player.SilenceTimeout = 3000

	c.Display.FgColor = "#ffffff"
	c.Display.BgColor = "#000000"
	c.Display.Font = "/usr/share/fonts/TTF/Hack-Regular.ttf"
	c.Display.FontSize = 32
	c.Display.WindowX = 410
	c.Display.WindowY = 0
	c.Display.WindowH = 1100
	c.Display.WindowW = 250

	return &c
}

func Load(path string) (*Config, error) {
	cfg := Default()

	if !util.FileExists(path) {
		return cfg, nil
	}

	if _, err := toml.DecodeFile(path, cfg); err != nil {
		return nil, err
	}

	return cfg, nil

}
